python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -1097,12 +1072,32 @@ def execute_asset_backfill_iteration(
 
         updated_backfill_data = updated_backfill.get_asset_backfill_data(asset_graph)
 
-        if backfill_is_complete(
-            backfill_id=backfill.backfill_id,
-            backfill_data=updated_backfill_data,
-            instance=instance,
+        partitions_without_materialization_status = (
+            updated_backfill_data.get_targeted_partitions_without_materialization_status()
+        )
+
+        asset_backfill_data_unchanged = (
+            updated_backfill_data.materialized_subset
+            == previous_asset_backfill_data.materialized_subset
+            and updated_backfill_data.requested_subset
+            == previous_asset_backfill_data.requested_subset
+            and updated_backfill_data.failed_and_downstream_subset
+            == previous_asset_backfill_data.failed_and_downstream_subset
+        )
+
+        if (
+            partitions_without_materialization_status.is_empty or asset_backfill_data_unchanged
+        ) and backfill_runs_are_complete(
             logger=logger,
+            backfill_runs=backfill_runs,
         ):
+            if not partitions_without_materialization_status.is_empty:
+                raise DagsterBackfillFailedError(
+                    "Backfill stopped producing runs, but the following partitions were never materialized or failed, "
+                    "possibly because the runs that were supposed to materialize them succeeded without creating the expected materializations: "
+                    f"{_asset_graph_subset_to_str(partitions_without_materialization_status, asset_graph)}."
+                )
```

Comment:
there's probably a less confusing way to structure this and separate out the two cases

case 1: backfill is not making progress, there are requested partitions not in a terminal state, and there are no more in progress or retrying runs - assume its stuck and fail the backfill

case 2: all requested partitions have been materialized or failed and there are no in progress runs - move to a successful or failed state depending on whether there are any failed partitions